File: app/src/main/java/com/mesh/client/transport/ChatTransport.kt
package com.mesh.client.transport

import android.util.Log
import com.mesh.client.crypto.CryptoManager
import com.mesh.client.data.EncryptedMessage
import com.mesh.client.network.WebRtcManager
import com.mesh.client.network.WebSocketService
import kotlinx.coroutines.launch

class ChatTransport(
    private val cryptoManager: CryptoManager,
    val webRtcManager: WebRtcManager,
    private val webSocketService: WebSocketService,
    private val myMeshId: String,
    private val outgoingDao: com.mesh.client.data.db.dao.OutgoingMessageDao,
    private val scope: kotlinx.coroutines.CoroutineScope
) : WebSocketService.Listener, WebRtcManager.Listener {

    interface MessageListener {
        fun onMessageReceived(fromMeshId: String, text: String, timestamp: Long, transportType: String)
        fun onMessageStatusChanged(peerId: String, isP2P: Boolean)
        
        // Protocol Events
        fun onInviteReceived(fromMeshId: String, inviteJson: String)
        fun onInviteAckReceived(fromMeshId: String, ackJson: String)
        fun onL2NotifyReceived(fromMeshId: String, notifyJson: String)
        
        fun onTransportError(message: String)
        fun onIncomingCall(peerId: String)
        fun onHangupReceived(peerId: String)
        fun onMessageDelivered(msgId: String) // New: for UI feedback
    }

    private val gson = com.google.gson.Gson()
    var messageListener: MessageListener? = null
    private var isConnected = false

    init {
        webRtcManager.listener = this
    }

    // --- Public Sending API ---
    
    fun sendMessage(peerId: String, plaintext: String) {
        val payload = com.mesh.client.data.ProtocolPayload("chat", plaintext)
        enqueueMessage(peerId, payload)
    }

    fun startCall(peerId: String) {
        // Ephemeral: try P2P first, then Server. No persistence.
        val payload = com.mesh.client.data.ProtocolPayload("call", "audio")
        sendEphemeral(peerId, payload)
        
        // Then establish WebRTC with audio track
        webRtcManager.startCall(peerId)
    }

    fun sendHangup(peerId: String) {
        val payload = com.mesh.client.data.ProtocolPayload("hangup", "ended")
        sendEphemeral(peerId, payload)
        webRtcManager.stopCall(peerId)
    }
    
    fun sendInvite(peerId: String, inviteJson: String) {
        // Invites should be persistent
        val payload = com.mesh.client.data.ProtocolPayload("invite", inviteJson)
        enqueueMessage(peerId, payload)
    }
    
    fun sendInviteAck(peerId: String, ackJson: String) {
        // ACKs can be ephemeral or persistent. Let's make invite ACKs persistent to be safe.
        val payload = com.mesh.client.data.ProtocolPayload("invite_ack", ackJson)
        enqueueMessage(peerId, payload)
    }

    private fun enqueueMessage(peerId: String, payload: com.mesh.client.data.ProtocolPayload) {
        scope.launch {
            try {
                val payloadJson = gson.toJson(payload)
                val encrypted = cryptoManager.encryptMessage(payloadJson, peerId)
                val encryptedJson = gson.toJson(encrypted)
                
                val entity = com.mesh.client.data.db.entities.OutgoingMessageEntity(
                    msgId = payload.msgId,
                    toPeerId = peerId,
                    encryptedJson = encryptedJson,
                    createdAt = System.currentTimeMillis()
                )
                outgoingDao.insert(entity)
                Log.d(TAG, "Enqueued message ${payload.msgId.take(8)} for $peerId")
                
                // Trigger sending immediately
                processOutbox()
            } catch (e: Exception) {
                Log.e(TAG, "Failed to enqueue message", e)
                messageListener?.onTransportError("Failed to encrypt/queue message")
            }
        }
    }
    
    fun processOutbox() {
        scope.launch {
            try {
                val pending = outgoingDao.getAllPending()
                if (pending.isEmpty()) return@launch
                
                Log.d(TAG, "Processing outbox: ${pending.size} messages")
                
                for (msg in pending) {
                    // Strategy: Server-First for reliability
                    // We send "server_message" via WebSocket.
                    
                    if (!isConnected) {
                        Log.d(TAG, "Cannot process outbox - WS disconnected")
                        break // Stop processing if no connection
                    }

                    try {
                        val encrypted = gson.fromJson(msg.encryptedJson, EncryptedMessage::class.java)
                        
                        // Send via Server (Reliable channel)
                        // Note: we pass msgId to the server envelope so we can get a persistent ACK
                        val success = webSocketService.sendEncryptedMessage(msg.toPeerId, encrypted, msg.msgId)
                        
                        if (success) {
                            Log.d(TAG, "Sent pending ${msg.msgId.take(8)} to server, waiting for release")
                            // We do NOT delete here. We wait for delivery_ack or offline_storage confirmation from server.
                        } else {
                            Log.w(TAG, "Failed to write ${msg.msgId.take(8)} to WS buffer")
                        }
                        
                    } catch (e: Exception) {
                        Log.e(TAG, "Error sending pending message ${msg.msgId}", e)
                    }
                }
            } catch (e: Exception) {
                Log.e(TAG, "Error processing outbox", e)
            }
        }
    }

    private fun sendEphemeral(peerId: String, payload: com.mesh.client.data.ProtocolPayload) {
        Log.i(TAG, "Sending ephemeral [${payload.type}] ID=${payload.msgId.take(8)} to $peerId")
        try {
            val payloadJson = gson.toJson(payload)
            val encrypted = cryptoManager.encryptMessage(payloadJson, peerId)

            // Try P2P first for ephemeral (latency sensitive)
            if (webRtcManager.isConnected(peerId)) {
                val sent = webRtcManager.sendP2PMessage(peerId, encrypted)
                if (!sent) webSocketService.sendEncryptedMessage(peerId, encrypted)
            } else {
                webSocketService.sendEncryptedMessage(peerId, encrypted)
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error sending ephemeral", e)
        }
    }

    // WebSocketService.Listener
    override fun onSignalingMessage(fromMeshId: String, type: String, payload: String?) {
        Log.d(TAG, "Signal [${type}] from ${fromMeshId.take(8)}")
        try {
            if (type == "error") return
            webRtcManager.handleSignaling(fromMeshId, type, payload)
        } catch (e: Exception) {
            Log.e(TAG, "Error handling signaling", e)
        }
    }

    override fun onEncryptedMessageReceived(fromMeshId: String, message: EncryptedMessage) {
        handleIncomingMessage(fromMeshId, message, "SERVER")
    }
    
    override fun onDeliveryAck(msgId: String, status: String) {
        handleServerAck(msgId, status)
    }
    
    override fun onError(message: String) {
        Log.e(TAG, "Transport Error: $message")
        messageListener?.onTransportError(message)
    }

    override fun onConnected() {
        Log.i(TAG, "Transport connected to Mesh Server")
        isConnected = true
        // Flush queue on reconnect
        processOutbox()
    }

    override fun onDisconnected() {
        Log.w(TAG, "Transport disconnected from Mesh Server")
        isConnected = false
    }

    // WebRtcManager.Listener
    override fun onP2PMessageReceived(fromMeshId: String, message: EncryptedMessage) {
        handleIncomingMessage(fromMeshId, message, "P2P")
    }

    override fun onP2PConnectionStateChange(peerId: String, isConnected: Boolean) {
        Log.i(TAG, "P2P State for ${peerId.take(8)}: $isConnected")
        messageListener?.onMessageStatusChanged(peerId, isConnected)
    }

    private fun handleIncomingMessage(fromMeshId: String, message: EncryptedMessage, source: String) {
        // Decrypt... same as before
        // But also check for delivery_ack inside? 
        // No, delivery_ack comes as a separate signaling/protocol message usually, 
        // OR as a payload inside an encrypted message?
        // Wait, main.py sends 'delivery_ack' as a personal message (JSON).
        // It is NOT encrypted as part of 'server_message' payload.
        // It is a top-level message type like 'webrtc_offer'.
        // So it comes via onSignalingMessage NO wait.
        
        // main.py calls manager.send_personal_message(ack).
        // ACK is { type: "delivery_ack", ... }
        // WebSocketService.kt sees "delivery_ack" in msgType.
        // It falls into 'else' -> onSignalingMessage.
        // So we need to handle "delivery_ack" in onSignalingMessage!
        
        try {
            val decryptedJson = cryptoManager.decryptMessage(message, fromMeshId)
            val payload = try {
                gson.fromJson(decryptedJson, com.mesh.client.data.ProtocolPayload::class.java)
            } catch (e: Exception) {
                com.mesh.client.data.ProtocolPayload("chat", decryptedJson)
            }

            Log.i(TAG, "Incoming [${payload.type}] ID=${payload.msgId.take(8)} from ${fromMeshId.take(8)}")
            
            // Auto-ACK for chat messages (Application Level ACK) purely for read receipt later
            // For now we rely on Transport ACK.

            when (payload.type) {
                "chat" -> messageListener?.onMessageReceived(fromMeshId, payload.content, message.timestamp, source.lowercase())
                "invite" -> messageListener?.onInviteReceived(fromMeshId, payload.content)
                "invite_ack" -> messageListener?.onInviteAckReceived(fromMeshId, payload.content)
                "l2_notify" -> messageListener?.onL2NotifyReceived(fromMeshId, payload.content)
                "call" -> messageListener?.onIncomingCall(fromMeshId)
                "hangup" -> {
                    webRtcManager.stopCall(fromMeshId)
                    messageListener?.onHangupReceived(fromMeshId)
                }
                else -> Log.w(TAG, "Unknown protocol type: ${payload.type}")
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to process incoming message", e)
        }
    }
    
    // We also need to handle the Server Delivery ACK which comes via onSignalingMessage/Protocol
    // See comment above.
    
    fun handleServerAck(msgId: String, status: String) {
        scope.launch {
            Log.d(TAG, "Server ACK for $msgId: $status")
            // If delivered or stored_offline, we consider it "sent" from client perspective.
            // We remove from outbox.
            outgoingDao.delete(msgId)
            messageListener?.onMessageDelivered(msgId)
        }
    }

    companion object {
        private const val TAG = "ChatTransport"
    }
}

File: web-client/src/core/network/ChatTransport.ts
import { CryptoManager } from '../crypto/CryptoManager';
import { WebRTCManager, WebRTCListener } from './WebRTCManager';
import { WebSocketService, WebSocketListener } from './WebSocketService';
import { EncryptedMessage, ProtocolPayload } from '../../types';

export interface ChatTransportListener {
    onMessageReceived(fromMeshId: string, text: string, timestamp: number): void;
    onMessageStatusChanged(peerId: string, isP2P: boolean): void;
    onInviteReceived(fromMeshId: string, inviteJson: string): void;
    onInviteAckReceived(fromMeshId: string, ackJson: string): void;
    onL2NotifyReceived(fromMeshId: string, notifyJson: string): void;
    onTransportError(message: string): void;
}

export class ChatTransport implements WebSocketListener, WebRTCListener {
    listener: ChatTransportListener | null = null;

    constructor(
        private cryptoManager: CryptoManager,
        private webRTCManager: WebRTCManager,
        private wsService: WebSocketService,
    ) {
        // Set up listeners
        this.wsService.listener = this;
        this.webRTCManager.listener = this;
    }

    async sendMessage(toPeerId: string, text: string): Promise<void> {
        const payload: ProtocolPayload = {
            type: 'chat',
            content: text,
        };
        await this.sendProtocolPayload(toPeerId, payload);
    }

    async sendInvite(toPeerId: string, inviteJson: string): Promise<void> {
        const payload: ProtocolPayload = {
            type: 'invite',
            content: inviteJson,
        };
        await this.sendProtocolPayload(toPeerId, payload);
    }

    async sendInviteAck(toPeerId: string, ackJson: string): Promise<void> {
        const payload: ProtocolPayload = {
            type: 'invite_ack',
            content: ackJson,
        };
        await this.sendProtocolPayload(toPeerId, payload);
    }

    private async sendProtocolPayload(peerId: string, payload: ProtocolPayload): Promise<void> {
        try {
            const payloadJson = JSON.stringify(payload);
            const encrypted = await this.cryptoManager.encryptMessage(payloadJson, peerId);

            // Try P2P first
            if (this.webRTCManager.isConnected(peerId)) {
                console.log(`Sending ${payload.type} via P2P to ${peerId}`);
                const sent = this.webRTCManager.sendP2PMessage(peerId, encrypted);

                if (!sent) {
                    console.warn('P2P send failed, falling back to server');
                    this.wsService.sendEncryptedMessage(peerId, encrypted);
                }
            } else {
                console.log(`Sending ${payload.type} via server to ${peerId} (P2P not connected)`);
                this.wsService.sendEncryptedMessage(peerId, encrypted);

                // Option: We could try to establish P2P for *future* messages,
                // but doing it on every message sent might spam offers if the peer is truly offline.
                // Better strategy: Only try to connect if we haven't tried recently or let a background process handle it.
                // For now, let's NOT aggressively connect here to avoid the 'recipient_offline' error loop if they are offline.
                // If they come online, they will send us an offer or we will eventually try again via a separate mechanism.
                // this.webRTCManager.connectToPeer(peerId); 
            }
        } catch (error) {
            console.error('Error sending protocol message:', error);
            this.listener?.onTransportError(`Failed to send message: ${error}`);
        }
    }

    // WebSocketListener implementation
    onSignalingMessage(fromMeshId: string, type: string, payload?: string): void {
        if (type === 'error') {
            this.listener?.onTransportError(payload || 'Unknown server error');
            return;
        }

        this.webRTCManager.handleSignaling(fromMeshId, type, payload);
    }

    onEncryptedMessageReceived(fromMeshId: string, message: EncryptedMessage): void {
        this.handleIncomingMessage(fromMeshId, message);
    }

    onConnected(): void {
        console.log('Transport connected');
    }

    onDisconnected(): void {
        console.log('Transport disconnected');
    }

    onError(message: string): void {
        this.listener?.onTransportError(message);
    }

    // WebRTCListener implementation
    onP2PMessageReceived(fromMeshId: string, message: EncryptedMessage): void {
        this.handleIncomingMessage(fromMeshId, message);
    }

    onP2PConnectionStateChange(peerId: string, isConnected: boolean): void {
        this.listener?.onMessageStatusChanged(peerId, isConnected);
    }

    private async handleIncomingMessage(fromMeshId: string, message: EncryptedMessage): Promise<void> {
        try {
            // STRATEGY: "Upgrade on Incoming"
            // If we receive a message via Server (which reaches here), it implies the peer is Online.
            // Check if we already have a P2P connection. If not, trigger a connection attempt.
            // We use a debounced check or simply rely on the manager's check (it handles duplicates).
            if (!this.webRTCManager.isConnected(fromMeshId)) {
                console.log(`[ChatTransport] Received message via Server from ${fromMeshId}. Triggering P2P upgrade.`);
                // We assume 'connectToPeer' safely handles multiple calls (check internal map).
                this.webRTCManager.connectToPeer(fromMeshId);
            }

            const decryptedJson = await this.cryptoManager.decryptMessage(message, fromMeshId);

            // Try to parse as ProtocolPayload
            try {
                const payload = JSON.parse(decryptedJson) as ProtocolPayload;

                switch (payload.type) {
                    case 'chat':
                        this.listener?.onMessageReceived(fromMeshId, payload.content, message.timestamp);
                        break;
                    case 'invite':
                        this.listener?.onInviteReceived(fromMeshId, payload.content);
                        break;
                    case 'invite_ack':
                        this.listener?.onInviteAckReceived(fromMeshId, payload.content);
                        break;
                    case 'l2_notify':
                        this.listener?.onL2NotifyReceived(fromMeshId, payload.content);
                        break;
                    default:
                        console.warn(`Unknown protocol type: ${payload.type}`);
                }
            } catch {
                // Backward compatibility: assume plain text chat
                this.listener?.onMessageReceived(fromMeshId, decryptedJson, message.timestamp);
            }
        } catch (error) {
            console.error(`Decryption failed from ${fromMeshId}:`, error);
        }
    }
}
